src/pudl/output_diff/cli.py

```diff
@@ -0,0 +1,403 @@
+        return inspect(self.get_engine(fname))
+
+    def get_row_count(self, fname: str, table_name: str) -> int:
+        """Returns number of rows contained within the table."""
+        if not re.compile(r"^[a-zA-Z0-9_]+$").match(table_name):
+            raise ValueError("table_name is not SQL safe")
+        # Note that because of the above check, CWE-89 SQL injection issue
+        # is no longer possible below.
+        # Because we get the table names from sqlite metadata, these table
+        # names should be safe by definition also.
+        return (
+            self.get_engine(fname)
+            .execute(text(f"SELECT COUNT(*) FROM {table_name}"))  # nosec
+            .scalar()
+        )
+
+    def get_rows_as_df(self, fname: str, table_name: str) -> pd.DataFrame:
+        """Returns sqlite table contents as pandas DataFrame."""
+        con = self.get_engine(fname)
+        return pd.concat(
+            [df for df in pd.read_sql_table(table_name, con, chunksize=100_000)]
+        )
+
+
+class OutputPair:
+    """Represents pair of output.
+
+    Baseline and an experiment.
+    """
+
+    def __init__(
+        self, baseline: OutputBundle, experiment: OutputBundle, args: argparse.Namespace
+    ):
+        """Create new instance with given baseline and experiment."""
+        self.baseline = baseline
+        self.experiment = experiment
+        self.args = args
+
+    def compare_files(self) -> ListDiff:
+        """Produces GenericDiff that contains file basenames."""
+        baseline_files = set(self.baseline.list_files())
+        experiment_files = set(self.experiment.list_files())
+        return ListDiff.from_sets("File", baseline_files, experiment_files)
+
+    def compare_sqlite_databases(self, file_diff: ListDiff) -> list[CategoryDiff]:
+        """Generates series of structural diffs over sqlite databases.
+
+        Returns dictionary where keys are diff categories and values are GenericDiff
+        representing the differences at the given category.
+
+        The following categories of diffs are generated:
+        - filename/table_name, values are table names
+        - filename/table_name/schema, values are (col_name, col_type) tuples
+        - filename/table_name/rowcounts, values are [row_count]
+        - filename/table_name/rows, values are [distinct_row_count]
+        """
+        # TODO(rousik): using GenericDiff for rowcounts and rows are not ideal,
+        # perhaps better approach would be to have type-specific diffs that allow
+        # either numbers for left/right (for rowcount) or panda dataframes for rows.
+        # But let's work with the hacky thing for now.
+        output_diffs = []
+
+        def maybe_append_diff(diff: CategoryDiff):
+            """Appends diff to output_diffs if there are actual differences."""
+            if not diff.sides_equal():
+                output_diffs.append(diff)
+                diff.print_diff()
+
+        for fname in file_diff.both:
```

Comment:
This seems like a reasonable way to encode "you can only diff table names between sqlite files that actually match", "you can only diff table structures between tables that are named the same", etc. 

If we were to add more types of diffs in the future, these dependencies might get a bit harder to parse out - maybe at that point we could switch to a pattern like

```python
        categories = ["file", "tablename", "schema", "row_count", "row"]
        category_funcs = {"file": ...}

        category_diffs = {}
        for category in categories:
            diff = category_funcs[category](category_diffs) # in each function, we'd look at the previous outputs to determine what to run diff calcs on
            category_diffs[category] = diff

        return [category_diffs[category] for category in categories]
```

We'd then have to make sure the `categories` list is in a specific order (bleh), but for each category func it would be clear which other categories it depends on.

If we wanted to go farther we could explicitly keep track of the deps in a dictionary, and then build a DAG, but that seems likely overkill :)